Projects/Server/Network/RingPipe.cs

```diff
@@ -0,0 +1,271 @@
+using System;
+using System.Runtime.InteropServices;
+
+namespace Server.Network;
+
+public partial class RingPipe : IDisposable
+{
+    public class PipeWriter
+    {
+        private readonly RingPipe _pipe;
+
+        internal PipeWriter(RingPipe pipe) => _pipe = pipe;
+
+        public unsafe Span<byte> TryGetMemory()
+        {
+            var read = _pipe._readIdx;
+            var write = _pipe._writeIdx;
+
+            uint sz;
+            if (read <= write)
+            {
+                sz = _pipe.Size - write + read - 1;
+            }
+            else
+            {
+                sz = read - write - 1;
+            }
+
+            return new Span<byte>((void*)(_pipe._buffer + write), (int)sz);
+        }
+
+        public void Advance(uint count)
+        {
+            var read = _pipe._readIdx;
+            var write = _pipe._writeIdx;
+
+            if (count == 0)
+            {
+                return;
+            }
+
+            if (count > _pipe.Size - 1)
+            {
+                throw new InvalidOperationException();
```

Comment:
Worth a specific exception? Maybe not...

Reply:
yeah, lots of this needs refining, good call out!